test/integration/master_benchmark_test.go

```diff
@@ -0,0 +1,92 @@
+// +build,!no-etcd !integration
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+	"io/ioutil"
+	"testing"
+)
+
+// If you need to start an etcd instance by hand, you also need to insert a key
+// for this check to pass (*any* key will do, eg:
+//curl -L http://127.0.0.1:4001/v2/keys/message -XPUT -d value="Hello world").
+func init() {
+	requireEtcd()
+}
+
+// Benchmark rc lifecycle
+func BenchmarkControllerCreate(b *testing.B) {
+	b.StopTimer()
+	m := NewMasterComponents()
+	defer m.Stop()
+
+	data, err := ioutil.ReadFile("benchmark-controller.json")
+	if err != nil {
+		glog.Fatalf("Unexpected error: %v", err)
+	}
+	glog.Infof("Starting controller benchmark with N=%d", b.N)
+	errs := []error{}
+	b.StartTimer()
+	for n := 0; n < b.N; n++ {
+		// StartRc waits around for the replicas to get created
+		rc, err := m.StartRc(string(data), uint(n), string(util.NewUUID()))
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		// StopRc waits around for the replicas to get deleted
+		err = m.StopRc(rc)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	b.StopTimer()
+	for err := range errs {
+		glog.Errorf("Unexpected error: %v", err)
+	}
+}
+
+// Benchmark pod listing
+func BenchmarkPodList(b *testing.B) {
+	b.StopTimer()
+	m := NewMasterComponents()
+	defer m.Stop()
+
+	data, err := ioutil.ReadFile("benchmark-controller.json")
+	if err != nil {
+		glog.Fatalf("Unexpected error: %v", err)
+	}
+	m.StartRc(string(data), uint(b.N), string(util.NewUUID()))
+
+	glog.Infof("Starting pod listing benchmark with N=%d", b.N)
+	errs := []error{}
+	b.StartTimer()
+	for n := 0; n < b.N; n++ {
```

Comment:
In this test, you're scaling the number of pods with b.N above. Then you're timing how long it takes to list them here. If you also list N times, then the running time will be N^2. If you think a single list isn't representative (and it may well not be), then you could do e.g. 100 lists and see if that makes the timings more stable.